### DISCUSSION TITLE
Practice Problem - Design an ATM

### DISCUSSION THREAD

**Discussion by kanmaytacker** [kanmaytacker/design-questions#11] [General]
What?
Create the design for an ATM.

Why?
Real world entity problem.
Popular in initial design rounds and traditional organisations.

How?
Follow the instructions [here](https://github.com/kanmaytacker/design-questions/blob/master/practice-problems/01-design-atm-worksheet.md).

**Comment by sastepankaj** [comment]
Hi,
Please find my class diagram for ATM. I have considered following as assumptions:
1. ATM will allow only one operation at a time and for a single user account.
2. ATM will communicate with Bank for transactions and Bank will take care of concurrency.

Question: I have shown relations such as "User interacts with ATM", "ATM communicates with Bank" etc using Association. Is this right way to depict the relation?

![Design ATM](https://user-images.githubusercontent.com/13418342/202782163-0e22b786-b075-4a9c-82b3-a6a60a785922.jpg)

**Comment by Faizahmadfaiz** [comment]
### Class Diagram

```mermaid
classDiagram
class Bank {
    -String name
    -Customer[] customers
}

class BankAccount {
    -String accountNumber
    -String debitCardNumber
    -Double balance
    +Credit(double)
    +debit(double)
}
class Atm {
    -String name
    -String address
    -Double totalAmount
    -Integer TwoThousandNotes
    -Integer FiveHundredNotes
    -Integer TwoHundredNotes
    -Integer HundredNotes
    +dispenseCash()
    +depositCash()
    +depositCheck()
    +checkBalance()
    +validateUser(User)
    +cancelTransaction()
}

class CardReader {
    +read()
}

class KeyPad {
    +readKey()
}

class Screen {
    +display()
}


class Printer {
    +print()
}

class Receipt {
    -Date issueDate
    -Double debitAmount
    -Customer customer 
    +print()
}

class User {
    -String name
}

class Customer {
    -BankAccount account
    -Integer atmPin
    +checkBalance()
    +depositCash()
    +depositCheck()
    +withDrawCash()
    +withDrawCheck()
}

class Admin {
    +checkLogs()
}

User <-- Customer
User <-- Admin
Customer --o BankAccount
Bank --o Customer
Bank --* Atm
Atm --* CardReader
Atm --* KeyPad
Atm --* Screen
Atm --* Printer

Atm --o Admin
```
